File: leetcode/everyday/easy_206_reverseList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
    // cur: 当前节点; pre: 上一个节点
    let cur = head, pre = null;
    while (cur) {
        // 记录当前节点后面节点, 防止后面被删除
        let tmp = cur.next
        // 当前节点的下一个节点指向上一个接地那
        cur.next = pre;
        // 当前节点设置为上一个节点
        pre = cur;
        // 将后续节点设置为当前节点
        cur = tmp;
    }
    return pre;
};

function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val);
    this.next = (next === undefined ? null : next)
}

let head = [1, 2, 3, 4, 5];
let first = new ListNode(-1), pre;
for (let i = 0; i < head.length; i++) {
    let node = new ListNode(head[i]);
    if (i == 0) {
        first.next = node;
    } else {
        pre.next = node;
    }
    pre = node;
}

reverseList(first.next)

File: list/easy_206_reverseList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        pre, cur = None, head
        while cur:
            next = cur.next
            cur.next = pre
            pre = cur
            cur = next
        return pre


l1 = ListNode(1)
l2 = ListNode(2, l1)
l3 = ListNode(3, l2)
l4 = ListNode(4, l3)
l5 = ListNode(5, l4)

#    head -> 5->4->3->2->1
# 1  head -> 4->5->3->2->1
# 2  head ->
s = Solution()
print('res: ', s.reverseList(l5))
